### ISSUE TITLE
Option to prevent duplicate images from webcam when using ffmpeg Reader.get_next_data()?

### ISSUE THREAD

**Issue by dennisvang** [imageio/imageio#287]
Below is a feature proposal, but I would be happy to turn it into a pull request.

Problem
----------

As far as I can see, when used with a webcam stream, `FfmpegFormat.Reader.get_next_data()` always returns an `Image`, regardless of whether-or-not a _new_ image is actually available from the webcam. Thus, the image returned may be either new or a duplicate of the previous one. 

This is probably intentional, but, if the user is unaware and expects a kind of blocking behavior (as I was/did initially...), it may lead to unexpected results. 

For example, in the code below, without additional timing control, we will get many unnecessary calls to `update_screen` and `append_data`, with duplicate frames in the resulting video file (assuming the loop frequency is higher than the camera framerate).

    while True:
        new_image = reader.get_next_data()
        update_screen(new_image)
        writer.append_data(new_image)

To fix this, the user would need to check for duplicate images or implement some other form of timing control.

Relevance
-----------

On the one hand, I can see why such a check for duplicates may be considered an implementation detail, but, on the other hand, I guess there are very few use cases where the user will _not_ benefit from a built-in (but optional) check for duplicates.

Proposed solution
--------------------

A convenient way to implement a check for duplicates in the ffmpeg `Reader.get_next_data()` would be to provide the _option_ to return `None` in case the image is a duplicate of the previous image (e.g. the `only_new` argument below). A user can then time his/her loop using a simple truth test:
    
    new_image = reader.get_next_data(only_new=True)
    if new_image is not None:
        update_screen(new_image)
        writer.append_data(new_image)

Setting `only_new=False` by default would ensure backward compatibility.

To reduce the performance penalty resulting from the equality test, we can test a limited number of pixels, e.g. let `only_new` take a (percentage) value instead of a boolean.

Alternatives
-------------
Maybe similar functionality is already available but I haven't found it yet?

Possible alternatives would be:

- ignore this issue and just let the user deal with it
- let the `get_next_data()` method block until a new frame is available
- use an event-driven approach, e.g. a `data_available` event (not sure if this can be implemented via ffmpeg and/or subprocess)

**Comment by almarklein** [comment]
This looks like a useful feature indeed. Note that the ffmpeg plugin is aware of when a duplicate frame is returned:
```py
            if index == self._pos:
                return self._lastread, {}
```

Perhaps we can add a sort of mode parameter so that the user can switch between the current behavior (allow duplicates), returning `None` if no data is available, and block.

**Comment by dennisvang** [comment]
Right, now I see, no need for an equality test after all. That makes life easier. A `mode` parameter sounds good. I'll look into it.

**Comment by dennisvang** [comment]
@almarklein : I suppose the obvious place to implement this would be `FfmpegFormat.Reader._get_data()`?

In that case I see one issue: returning `None` would cause `Image()` to fail, because it expects a Numpy-array. This could be tackled in `Format.Reader.get_data()` or in `Image.__new__()`, but I guess it would be preferable not to mess around with those, and limit any edit to the ffmpeg plugin?

Perhaps, instead of returning `None`, we could add an item to the return-dict (image metadata) from `FfmpegFormat.Reader._get_data()`? This way the `mode` parameter would not be necessary (unless we also want to add a blocking mode). This seems like the most minimally-invasive solution to me. For example:

```
if index == self._pos:
    return self._lastread, {'old'=True}
```
Is it okay to use the meta dict for this purpose?

**Comment by dennisvang** [comment]
@almarklein : Upon closer inspection it seems that `index == self._pos` is never True when using `get_next_data()`, because `get_next_data()` always increments `index` (is that correct?). 

Thus, `_read_frame()` is called every time `get_next_data()` is called, and the above will not work. We still end up with duplicate frames because `_frame_catcher.get_frame()` just returns the old `_frame_catcher._frame`. 

The `FrameCatcher` does know when a new frame is available, so I guess some flag and/or `threading.Condition` should be added there.

**Comment by almarklein** [comment]
Closed by #288
